ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,250 @@
+producing a container-image at the end.
+
+![source-upload sequence diagram](assets/0005-source-upload-sequence-diagram.png)
+
+To explain the sequence of events in more detail, please consider the following topics.
+
+#### Prepare _(Step 1 to 3)_
+
+The preparation consists on declaring a BuildRun object _(1)_ with configuration for the
+Build-Controller, to instead of appending the regular "`git clone`" step, to wait for the user input
+_(2)_. The waiting process will have a configurable timeout.
+
+On the CLI, it will be watching for the creation of pods using a `LabelSelector` _(3)_,  which
+narrows the search criteria to only find the pod triggered by step 1.
+
+#### Upload _(Step 4 to 7)_
+
+Shipwright-Controller sets up a Tekton Workspace named "source", and therefore the build-pod
+containers have a `/workspace/source` volume mounted, this volume will receive the uploaded data. The
+build-pod waits _(4)_ for the upload by keeping the entrypoint process active until the end of the
+upload.
+
+The CLI knows in which state _(5)_ the build-pod is at, since the data upload can only happen when
+the build-pod is running. Given the right conditions, it streams the local data to the cluster _(6)_.
+
+Once the streaming is finished, the CLI notifies the waiting process the upload is done _(7)_, so
+inside the build-pod the execution of subsequent containers can take place.
+
+#### Build _(Step 8 and 9)_
+
+When the build steps are being executed _(8)_ in the build-pod, the developer will have the execution
+logs being shown in the command-line _(9)_. The developer can interactively watch the container build
+execution and possible errors that might surface.
+
+At the end, when the build-pod is completed successfully _(5)_, the CLI shows the meta information
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][client-go-watch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][event-loop-example].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][event-loop-usage].
+
+### Upload Mechanism
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][kubectl-cp]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
```

Comment:
some reiteration that this is part of the "pluggable mechanism" aspect of the story could be useful here, where we let users opt it

hence, if they are comfortable giving a developer the requisite permissions in the namespace in question to do `kubectl cp`, great.

Otherwise, if they are more comfortable with the requisite permissions need to employ the image registry techniques cited in https://github.com/shipwright-io/build/blob/main/docs/proposals/enable-local-source-code-support.md, great.

Reply:
I'm wording this on the beginning of the section, please consider.

Reply:
OK yep see it above.  I'm going to add a comment there, but feel free to mark this thread resolved.